util/lock.go

```diff
@@ -0,0 +1,173 @@
+	_, err = lock.TryLock()
+	if err != nil {
+		return nil, err
+	}
+	go lock.connCheck()
+	return lock, nil
+}
+
+func getConn(pool *sql.DB) (*sql.Conn, error) {
+	lockConn, err := pool.Conn(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("error getting DB connection: %v", err)
+	}
+	return lockConn, nil
+}
+
+func (l *PgAdvisoryLock) Id() string {
+	return strconv.Itoa(l.groupLockId)
+}
+
+func (l *PgAdvisoryLock) BecomeLeader() (bool, error) {
+	return l.TryLock()
+}
+
+func (l *PgAdvisoryLock) IsLeader() (bool, error) {
+	return l.Locked(), nil
+}
+
+func (l *PgAdvisoryLock) Resign() error {
+	return l.Release()
+}
+
+func (l *PgAdvisoryLock) TryLock() (bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+	if l.closed {
+		return false, fmt.Errorf("lock is closed. please create a new lock")
+	}
+	if l.obtained {
+		return l.obtained, nil
+	}
+	err := checkConnection(l.conn)
+	if err != nil {
+		conn, err := getConn(l.connPool)
+		if err != nil {
+			return false, err
+		}
+		l.conn = conn
+	}
+	rows, err := l.conn.QueryContext(context.Background(), "SELECT pg_try_advisory_lock($1)", l.groupLockId)
+	if err != nil {
+		return false, err
+	}
+	defer rows.Close()
+	if !rows.Next() {
+		return false, fmt.Errorf("error while trying to read response rows from `pg_try_advisory_lock` function")
+	}
+	var success bool
+	if err := rows.Scan(&success); err != nil {
+		return false, err
+	}
+	l.obtained = success
+	if l.obtained {
+		log.Debug("msg", fmt.Sprintf("Lock obtained for group id %d", l.groupLockId))
+	}
+	return success, nil
+}
+
+func (l *PgAdvisoryLock) Locked() bool {
+	l.mutex.RLock()
+	defer l.mutex.RUnlock()
+	return l.obtained
+}
+
+func (l *PgAdvisoryLock) Release() error {
+	l.mutex.Lock()
+	if !l.obtained {
+		return fmt.Errorf("can't release while not holding the lock")
+	}
+	defer l.mutex.Unlock()
+	rows, err := l.conn.QueryContext(context.Background(), "SELECT pg_advisory_unlock($1)", l.groupLockId)
+	if err != nil {
+		return err
+	}
+	rows.Next()
+	var success bool
+	if err := rows.Scan(&success); err != nil {
+		return err
+	}
+	if !success {
+		return fmt.Errorf("failed to release a lock with group lock id: %v", l.groupLockId)
+	}
+	rows.Close()
+	l.conn.Close()
+	l.obtained = false
+	return nil
+}
+
+func (l *PgAdvisoryLock) Close() error {
+	err := l.Release()
+	l.quit <- true
+	return err
+}
+
+func checkConnection(conn *sql.Conn) error {
+	_, err := conn.ExecContext(context.Background(), "SELECT 1")
+	if err != nil {
+		return fmt.Errorf("invalid connection: %v", err)
+	}
+	return nil
+}
+
+func (l *PgAdvisoryLock) connCheck() {
+	for {
+		select {
+		case <-l.ticker.C:
+			l.mutex.RLock()
+			err := checkConnection(l.conn)
+			l.mutex.RUnlock()
+			if err != nil {
+				l.mutex.Lock()
+				l.obtained = false
+				l.mutex.Unlock()
+			}
+		case <-l.quit:
+			l.mutex.Lock()
+			l.closed = true
```

Comment:
When does this release the advisory lock if we're the leader?

Reply:
The lock is being released on `Close()`. This method is called internally